internal/jobs/jobobject.go

```diff
@@ -0,0 +1,376 @@
+		} else if limits.cpuWeight != 0 {
+			if limits.cpuWeight < processorWeightMin || limits.cpuWeight > processorWeightMax {
+				return fmt.Errorf("processor weight value of `%d` is invalid", limits.cpuWeight)
+			}
+			cpuInfo.ControlFlags = winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_WEIGHT_BASED
+			cpuInfo.Value = calculateJobCPUWeight(limits.cpuWeight)
+		} else if limits.cpuRate != 0 {
+			if limits.cpuRate < processorLimitMin || limits.cpuRate > processorLimitMax {
+				return fmt.Errorf("processor rate of `%d` is invalid", limits.cpuRate)
+			}
+			cpuInfo.ControlFlags = winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP
+			cpuInfo.Value = limits.cpuRate
+		}
+		_, err := windows.SetInformationJobObject(job.jobHandle, windows.JobObjectCpuRateControlInformation, uintptr(unsafe.Pointer(&cpuInfo)), uint32(unsafe.Sizeof(cpuInfo)))
+		if err != nil {
+			return fmt.Errorf("failed to set cpu limit info on job object: %s", err)
+		}
+	}
+
+	if limits.maxBandwidth != 0 || limits.maxIops != 0 {
+		ioInfo := winapi.JOBOBJECT_IO_RATE_CONTROL_INFORMATION{
+			ControlFlags: winapi.JOB_OBJECT_IO_RATE_CONTROL_ENABLE,
+		}
+		if limits.maxBandwidth != 0 {
+			ioInfo.MaxBandwidth = limits.maxBandwidth
+		}
+		if limits.maxIops != 0 {
+			ioInfo.MaxIops = limits.maxIops
+		}
+		_, err := winapi.SetIoRateControlInformationJobObject(job.jobHandle, &ioInfo)
+		if err != nil {
+			return fmt.Errorf("failed to set IO limit info on job object: %s", err)
+		}
+	}
+	return nil
+}
+
+// CreateJobObject creates a job object, attaches an IO completion port to use
+// for notifications and then returns an object with the corresponding handles.
+func CreateJobObject(name string) (*JobObject, error) {
+	jobHandle, err := windows.CreateJobObject(nil, windows.StringToUTF16Ptr(name))
+	if err != nil {
+		return nil, err
+	}
+
+	iocpChan := make(chan IOCPNotif, 10)
+	jobMap[uint32(jobHandle)] = iocpChan
+	// If the IOCP we'll be using to receive messages for all jobs hasn't been
+	// created, create it and start polling.
+	var initIOErr error
+	ioInitOnce.Do(func() {
+		h, err := windows.CreateIoCompletionPort(windows.InvalidHandle, 0, 0, 0xffffffff)
+		// If this fails we don't have any way of monitoring when a job container has
+		// shutdown fully. This is run once via sync.Once.Do so we can't just bubble up an error
+		// here.
+		if err != nil {
+			initIOErr = err
+			return
+		}
+		ioCompletionPort = h
+		go pollIOCP(h)
+	})
+
+	if initIOErr != nil {
+		return nil, err
+	}
+
+	if _, err = attachIOCP(jobHandle, ioCompletionPort); err != nil {
+		windows.Close(jobHandle)
+		return nil, err
+	}
+	return &JobObject{
+		jobHandle,
+		iocpChan,
+		sync.RWMutex{},
+	}, nil
+}
+
+// IOCPChan returns the channel that the job object will receive IO completion port
+// notifications on.
+func (job *JobObject) IOCPChan() chan IOCPNotif {
```

Comment:
I'd like to see if we can present a nicer interface than this. Parsing out the details of the IOCP message should be done in this package, and we should treat the fact that we are using an IOCP as an implementation detail.

Reply:
Can we define nicer types for the different messages, and pass a `interface{}` over the channel instead?

Reply:
Trying to wrap my head around your vision here so bear with me :). Do you mean locally in the jobobject package we can see that we got x,y,z etc. iocp message and then define types that we can return to signify something that we care about in the jobcontainers package but that isn't tied to the notion of it coming from an iocp? E.g. we can see in package jobobject that we received `JOB_OBJECT_MSG_ACTIVE_PROCESS_ZERO` and what we'd actually send over would be some object like `jobobject.Completed`?

Reply:
Mostly unclear as to what we'd send over type wise. All we actually parse in the implementation so far is just `JOB_OBJECT_MSG_ACTIVE_PROCESS_ZERO` but we could add limit messages to this list for debugging (or if we ever swap to killing the container if it reaches a limit this could be re-used). Not sure what the types you're envisioning for the different messages would look like

Reply:
Basically, yes. We should abstract away:
- The fact that these messages are delivered via IOCP
- The underlying representation of these messages

As you said, for a `JOB_OBJECT_MSG_ACTIVE_PROCESS_ZERO`, we can return some sort of `jobobject.Completed` or `jobobject.AllProcessesTerminated` type.

For a `JOB_OBJECT_MSG_NOTIFICATION_LIMIT`, we can query the limit that was exceeded and return a `jobobject.LimitExceeded` type that has fields indicating things like the offending process, and the details of the limit that was hit.

Reply:
Ok sweet, I like the sound of that a lot